doc/metrics.md

```diff
@@ -0,0 +1,286 @@
+# Metrics
+
+Both the Java client and the Clojure client have [Dropwizard
+Metrics](http://metrics.dropwizard.io/3.1.0/) support - they both accept as an
+option a `MetricRegistry` to which they will register http metrics for each
+request, as well as metrics for an metric-ids specified in the request options.
+
+For using metrics with either the Java client and the Clojure client you must
+already have created a Dropwizard `MetricRegistry`.
+
+- [Metrics prefix](#metrics-prefix)
+- [Types of metrics](#types-of-metrics)
+- [Getting back metrics](#getting-back-metrics)
+- [Java API](#java-api)
+  - [Creating a client with metrics](#creating-a-client-with-metrics)
+  - [Setting a metric-id](#setting-a-metric-id)
+  - [Filtering metrics](#filtering-metrics)
+    - [Filtering by URL](#filtering-by-url)
+    - [Filtering by URL and method](#filtering-by-url-and-method)
+    - [Filtering by metric-id](#filtering-by-metric-id)
+- [Clojure API](#clojure-api)
+  - [Creating a client with metrics](#creating-a-client-with-metrics-1)
+  - [Setting a metric-id](#setting-a-metric-id-1)
+  - [Filtering metrics](#filtering-metrics-1)
+    - [Filtering by URL](#filtering-by-url-1)
+    - [Filtering by URL and method](#filtering-by-url-and-method-1)
+    - [Filtering by metric-id](#filtering-by-metric-id-1)
+
+
+## Metrics prefix
+
+All http metrics are prefixed with `puppetlabs.http-client.experimental`.
+
+## Types of metrics
+
+There are two types of metrics: *full response* and *initial response*. Full
+response metrics stop when all bytes of the response have been read by the
+client. Initial response metrics stop when the first byte of the response has
+been received by the client. Full response metrics are suffixed with
+`full-response`. Initial response metric support has not yet been
+implementedare suffixed with `full-response`. Initial response metric support
+has not yet been implemented.
+
+There are three categories of metrics: `url` metrics, `url-and-method`
+metrics, and `metric-id` metrics. `url` and `url-and-method` metrics are
+created automatically for each request; `metric-id` metrics are only created
+for requests that have a `metric-id` request option specified.
+
+Each http request will have a metric name created for its url (stripped of
+query strings and url fragments), as well as for the url + method name. `url`
+metrics have `with-url` after the prefix, followed by the url.
+`url-and-method` metrics have `with-url-and-method` after the prefix, followed
+by the url and then the capitalized HTTP method.
+
+So, for example,  a `GET` request to `http://foobar.com` would create a metric
+`puppetlabs.http-client.experimental.with-url.http://foobar.com.full-response`
+and a metric
+`puppetlabs.http-client.experimental.with-url-and-method.http://foobar.com.GET.full-response`.
+
+It is also possible to give a request a `metric-id`. The `metric-id` is an
+array of strings.  For each element in the array, a metric name will be
```

Comment:
I think I might leave this as is, although I'm waffling on it. In the Clojure API they can be queried using keywords, but they're actually represented (and returned) as strings.